cpp/src/io/parquet/writer_impl.cu

```diff
@@ -1412,10 +1436,22 @@ void writer::impl::write(table_view const& table, std::vector<partition_info> co
   // iteratively reduce this value if the largest fragment exceeds the max page size limit (we
   // ideally want the page size to be below 1MB so as to have enough pages to get good
   // compression/decompression performance).
-  // If using the default fragment size, scale it up or down depending on the requested page size.
+  // If using the default fragment size, adapt fragment size so that page size guarantees are met.
   if (max_page_fragment_size_ == cudf::io::default_max_page_fragment_size) {
     max_page_fragment_size_ = (cudf::io::default_max_page_fragment_size * max_page_size_bytes) /
                               cudf::io::default_max_page_size_bytes;
+
+    if (table.num_rows() > 0) {
+      std::for_each(
+        table.begin(), table.end(), [this, num_rows = table.num_rows()](auto const& column) {
+          auto const avg_len = column_size(column, stream) / num_rows;
+
+          if (avg_len > 0) {
+            size_type frag_size     = max_page_size_bytes / avg_len;
```

Comment:
is this too large? IIUC, `max_page_size_bytes / avg_len` is the average number of rows in each page. That means that any deviation in size between rows would cause us to overshoot the max page size.

Reply:
Yes, avg will tend to overshoot.  But with the deeply nested cases, it we overshoot anyway, now just by much less. :sweat_smile: I could try max row length perhaps, but that will be trickier to calculate than total size for the nested case.

Reply:
I tried to suggest a better option, but anything small enough to get really precise page sizes in the largest column would significantly degrade performance for other columns. 
Current implementation is actually a good compromise.